android-core/src/main/java/com/mparticle/internal/DeviceAttributes.kt

```diff
@@ -0,0 +1,343 @@
+     *
+     *
+     * This contains logic that MUST only be called once per app run.
+     *
+     * @param appContext the application context
+     * @return a JSONObject of device-specific attributes
+     */
+    fun getStaticDeviceInfo(appContext: Context): JSONObject {
+        val attributes = JSONObject()
+
+        try {
+            // device/OS attributes
+            attributes.put(MessageKey.BUILD_ID, Build.ID)
+            attributes.put(MessageKey.BRAND, Build.BRAND)
+            attributes.put(MessageKey.PRODUCT, Build.PRODUCT)
+            attributes.put(MessageKey.DEVICE, Build.DEVICE)
+            attributes.put(MessageKey.MANUFACTURER, Build.MANUFACTURER)
+            attributes.put(MessageKey.PLATFORM, operatingSystemString)
+            attributes.put(MessageKey.OS_VERSION, Build.VERSION.SDK)
+            attributes.put(MessageKey.OS_VERSION_INT, Build.VERSION.SDK_INT)
+            attributes.put(MessageKey.MODEL, Build.MODEL)
+            attributes.put(MessageKey.RELEASE_VERSION, Build.VERSION.RELEASE)
+
+            val application = appContext as Application
+            // device ID
+            addAndroidId(attributes, application)
+
+            attributes.put(MessageKey.DEVICE_BLUETOOTH_ENABLED, MPUtility.isBluetoothEnabled(appContext))
+            attributes.put(MessageKey.DEVICE_BLUETOOTH_VERSION, MPUtility.getBluetoothVersion(appContext))
+            attributes.put(MessageKey.DEVICE_SUPPORTS_NFC, MPUtility.hasNfc(appContext))
+            attributes.put(MessageKey.DEVICE_SUPPORTS_TELEPHONY, MPUtility.hasTelephony(appContext))
+
+            val rootedObject = JSONObject()
+            rootedObject.put(MessageKey.DEVICE_ROOTED_CYDIA, MPUtility.isPhoneRooted())
+            attributes.put(MessageKey.DEVICE_ROOTED, rootedObject)
+
+            // screen height/width
+            val displayMetrics = appContext.getResources().displayMetrics
+            attributes.put(MessageKey.SCREEN_HEIGHT, displayMetrics.heightPixels)
+            attributes.put(MessageKey.SCREEN_WIDTH, displayMetrics.widthPixels)
+            attributes.put(MessageKey.SCREEN_DPI, displayMetrics.densityDpi)
+
+            // locales
+            val locale = Locale.getDefault()
+            attributes.put(MessageKey.DEVICE_COUNTRY, locale.displayCountry)
+            attributes.put(MessageKey.DEVICE_LOCALE_COUNTRY, locale.country)
+            attributes.put(MessageKey.DEVICE_LOCALE_LANGUAGE, locale.language)
+            attributes.put(MessageKey.DEVICE_TIMEZONE_NAME, MPUtility.getTimeZone())
+            attributes.put(MessageKey.TIMEZONE, TimeZone.getDefault().rawOffset / (1000 * 60 * 60))
+            // network
+            val telephonyManager = appContext
+                .getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
+            val phoneType = telephonyManager.phoneType
+            if (phoneType != TelephonyManager.PHONE_TYPE_NONE) {
```

Comment:
Nitpick, but seems like the extra variable assignment is unnecessary.

```suggestion
            if (telephonyManager.phoneType != TelephonyManager.PHONE_TYPE_NONE) {
```